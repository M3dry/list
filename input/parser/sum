(+ 4 1)

-- args: -i -p -f -e
-- expected stdout:
-- Ok(
--     Call(
--         Call {
--             func: Right(
--                 "+",
--             ),
--             params: [
--                 Literal(
--                     Int(
--                         false,
--                         4,
--                     ),
--                 ),
--                 Literal(
--                     Int(
--                         false,
--                         1,
--                     ),
--                 ),
--             ],
--         },
--     ),
-- )

