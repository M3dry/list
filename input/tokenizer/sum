(+ 4 1)

-- args: -i -f
-- expected stdout:
-- Ok(
--     Tokens(
--         [
--             ParenOpen,
--             Identifier(
--                 "+",
--             ),
--             Literal(
--                 Int(
--                     false,
--                     4,
--                 ),
--             ),
--             Literal(
--                 Int(
--                     false,
--                     1,
--                 ),
--             ),
--             ParenClose,
--         ],
--     ),
-- )

