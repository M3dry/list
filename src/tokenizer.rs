use std::{
    iter::Peekable,
    str::{Chars, FromStr},
};

#[derive(Debug, PartialEq, Eq)]
pub struct Tokens(pub(crate) Vec<Token>);

#[derive(Debug, PartialEq, Eq)]
pub(crate) enum Token {
    Literal(Literals),
    Keyword(Keywords),
    Type(BuiltinTypes),
    ParenOpen,
    ParenClose,
    BracketOpen,
    BracketClose,
    AngleBracketOpen,
    AngleBracketClose,
    CurlyOpen,
    CurlyClose,
    Identifier(String),
    Generic(String),
}

impl From<Vec<Token>> for Tokens {
    fn from(value: Vec<Token>) -> Self {
        Tokens(value)
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum Literals {
    Int(bool, u128),
    String(String),
    Char(char),
    Bool(bool),
}

impl ToString for Literals {
    fn to_string(&self) -> String {
        match self {
            Literals::Int(true, num) => format!("-{}", &num.to_string()),
            Literals::Int(false, num) => num.to_string(),
            Literals::String(str) => format!(r#""{str}""#),
            Literals::Char(char) => format!("'{char}'"),
            Literals::Bool(true) => "true".to_string(),
            Literals::Bool(false) => "false".to_string(),
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub(crate) enum Keywords {
    If,
    Match,
    Defun,
    Lambda,
    Let,
    Struct,
    Enum,
    Arrow,
}

#[derive(Debug, PartialEq, Eq)]
pub enum BuiltinTypes {
    U8,
    U16,
    U32,
    U64,
    U128,
    I8,
    I16,
    I32,
    I64,
    I128,
    String,
    Char,
    Bool,
}

impl ToString for BuiltinTypes {
    fn to_string(&self) -> String {
        match self {
            BuiltinTypes::U8 => "u8",
            BuiltinTypes::U16 => "u16",
            BuiltinTypes::U32 => "u32",
            BuiltinTypes::U64 => "u64",
            BuiltinTypes::U128 => "k128",
            BuiltinTypes::I8 => "i8",
            BuiltinTypes::I16 => "i16",
            BuiltinTypes::I32 => "i32",
            BuiltinTypes::I64 => "i64",
            BuiltinTypes::I128 => "i128",
            BuiltinTypes::String => "String",
            BuiltinTypes::Char => "char",
            BuiltinTypes::Bool => "bool",
        }
        .to_string()
    }
}

impl FromStr for Tokens {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut tokens = vec![];
        let mut chars = s.chars().peekable();

        'l: while let Some(char) = chars.next() {
            match char {
                '(' => tokens.push(Token::ParenOpen),
                ')' => tokens.push(Token::ParenClose),
                '[' => tokens.push(Token::BracketOpen),
                '<' => tokens.push(Token::AngleBracketOpen),
                '>' => tokens.push(Token::AngleBracketClose),
                ']' => tokens.push(Token::BracketClose),
                '{' => tokens.push(Token::CurlyOpen),
                '}' => tokens.push(Token::CurlyClose),
                '\'' => {
                    let ch = chars.next().unwrap();

                    if !matches!(chars.next(), Some('\'')) {
                        return Err("unfinished closing single quote");
                    }

                    tokens.push(Token::Literal(Literals::Char(ch)));
                }
                '"' => {
                    let mut literal = vec![chars.next().unwrap()];
                    let mut backslash = false;

                    while let Some(char) = chars.next() {
                        if char == '"' && !backslash {
                            break;
                        } else if char == '\\' && !backslash {
                            backslash = true
                        } else {
                            literal.push(char);
                            backslash = false
                        }
                    }

                    tokens.push(Token::Literal(Literals::String(
                        literal.into_iter().collect(),
                    )));
                }
                char => {
                    if char == '('
                        || char == ')'
                        || char == '['
                        || char == ']'
                        || char == '<'
                        || char == '>'
                        || char == ' '
                    {
                        continue;
                    }
                    let mut chs = vec![];

                    if char == '-' {
                        match check_arrow(&mut chars)? {
                            Triple::Token(token) => {
                                tokens.push(token);
                                continue 'l;
                            }
                            Triple::Char(char) => {
                                chs.push(char);
                            }
                            Triple::None => {}
                        }
                    } else {
                        chs.push(char)
                    }

                    while let Some(char) = chars.peek() {
                        if *char == '('
                            || *char == ')'
                            || *char == '['
                            || *char == ']'
                            || *char == '<'
                            || *char == '>'
                            || *char == ' '
                        {
                            break;
                        } else if *char == '-' {
                            let ch = chars.next().unwrap();
                            match check_arrow(&mut chars)? {
                                Triple::Token(token) => {
                                    tokens.append(&mut token_from_str(
                                        &chs.into_iter().collect::<String>()[..],
                                    ));
                                    tokens.push(token);
                                    continue 'l;
                                }
                                Triple::Char(char) => {
                                    chs.push(ch);
                                    chs.push(char);
                                }
                                Triple::None => {}
                            }
                        } else {
                            chs.push(chars.next().unwrap())
                        }
                    }

                    tokens.append(&mut token_from_str(
                        &chs.into_iter().collect::<String>()[..],
                    ));
                }
            }
        }

        Ok(Tokens::from(tokens))
    }
}

fn token_from_str(str: &str) -> Vec<Token> {
    match str {
        "if" => vec![Token::Keyword(Keywords::If)],
        "match" => vec![Token::Keyword(Keywords::Match)],
        "defun" => vec![Token::Keyword(Keywords::Defun)],
        "lambda" => vec![Token::Keyword(Keywords::Lambda)],
        "let" => vec![Token::Keyword(Keywords::Let)],
        "struct" => vec![Token::Keyword(Keywords::Struct)],
        "enum" => vec![Token::Keyword(Keywords::Enum)],
        "true" => vec![Token::Literal(Literals::Bool(true))],
        "false" => vec![Token::Literal(Literals::Bool(false))],
        "u8" => vec![Token::Type(BuiltinTypes::U8)],
        "u16" => vec![Token::Type(BuiltinTypes::U16)],
        "u32" => vec![Token::Type(BuiltinTypes::U32)],
        "u64" => vec![Token::Type(BuiltinTypes::U64)],
        "u128" => vec![Token::Type(BuiltinTypes::U128)],
        "i8" => vec![Token::Type(BuiltinTypes::I8)],
        "i16" => vec![Token::Type(BuiltinTypes::I16)],
        "i32" => vec![Token::Type(BuiltinTypes::I32)],
        "i64" => vec![Token::Type(BuiltinTypes::I64)],
        "i128" => vec![Token::Type(BuiltinTypes::I128)],
        "string" => vec![Token::Type(BuiltinTypes::String)],
        "char" => vec![Token::Type(BuiltinTypes::Char)],
        "bool" => vec![Token::Type(BuiltinTypes::Bool)],
        arrow if arrow.len() >= 2 && &arrow[..2] == "->" => {
            let mut ret = vec![Token::Keyword(Keywords::Arrow)];
            ret.append(&mut token_from_str(&str[2..]));
            ret
        }
        num if num.len() > 1 && &num[..1] == "-" && num[1..].parse::<u128>().is_ok() => {
            vec![Token::Literal(Literals::Int(
                true,
                num[1..].parse().unwrap(),
            ))]
        }
        num if num.parse::<u128>().is_ok() => {
            vec![Token::Literal(Literals::Int(false, num.parse().unwrap()))]
        }
        generic if generic.len() > 1 && &generic[..1] == ":" => {
            vec![Token::Generic(generic[1..].to_string())]
        }
        identifier
            if identifier.len() > 0
                && identifier.chars().all(|c| {
                    c.is_ascii_alphanumeric()
                        || [
                            '!', '@', '#', '$', '%', '^', '&', '*', '-', '_', '=', '+', ',', '.',
                            '<', '>', '?',
                        ]
                        .contains(&c)
                }) =>
        {
            vec![Token::Identifier(identifier.to_string())]
        }
        _ => vec![],
    }
}

enum Triple {
    Token(Token),
    Char(char),
    None,
}

fn check_arrow(chars: &mut Peekable<Chars>) -> Result<Triple, &'static str> {
    match chars.next() {
        Some('-') => {
            while !matches!(chars.next(), Some('\n') | None) {}
            Ok(Triple::None)
        }
        Some('>') => Ok(Triple::Token(Token::Keyword(Keywords::Arrow))),
        Some(char) => Ok(Triple::Char(char)),
        _ => Err("Expected more chars"),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! snapshot {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let contents = include_str!($path);
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("../testdata/tokenizer/");
                settings.bind(|| {
                    insta::assert_snapshot!(contents
                        .lines()
                        .filter_map(|line| if line != "" {
                            Some(format!("{line}\n{:#?}", Tokens::from_str(line)))
                        } else {
                            None
                        })
                        .collect::<Vec<String>>()
                        .join("\n\n"));
                });
            }
        };
    }

    snapshot!(test_calling, "../testdata/input/calling.lt");
    snapshot!(test_if, "../testdata/input/if.lt");
    snapshot!(test_match, "../testdata/input/match.lt");
    snapshot!(test_defun, "../testdata/input/defun.lt");
    snapshot!(test_lambda, "../testdata/input/lambda.lt");
    snapshot!(test_let, "../testdata/input/let.lt");
    snapshot!(test_struct, "../testdata/input/struct.lt");
    snapshot!(test_enum, "../testdata/input/enum.lt");
}
