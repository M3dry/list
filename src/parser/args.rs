use crate::tokenizer::{Keywords, Token};

use super::{
    error,
    r#type::{Generic, Type},
    Parser, ParserError, ParserErrorStack,
};

#[derive(Debug)]
pub(crate) enum Arg {
    Generic(Generic),
    Named(String, Type),
    Simple(String),
    SelfA(Type),
}

impl TryFrom<&mut Parser> for Arg {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let next = value.pop_front_err("Arg")?;
        let name = match next {
            Token::Identifier(iden) if &iden == "self" => return Ok(Arg::SelfA(Type::SelfA)),
            w @ Token::Ref => {
                value.tokens.push_front(w);
                let ret = error!(Type::try_from(&mut *value), "Arg")?;

                match &ret {
                    Type::Ref(_, t) | Type::RefMut(_, t) if matches!(**t, Type::SelfA) => {
                        return Ok(Self::SelfA(ret))
                    }
                    token => return Err(error!("Arg", format!("Expected self, got {token:#?}"))),
                }
            }
            w @ Token::Char(':') => {
                value.tokens.push_front(w);
                return Ok(Arg::Generic(error!(Generic::try_from(&mut *value), "Arg")?));
            }
            Token::Identifier(iden) => iden,
            _ => {
                return Err(error!(
                    "Arg name",
                    format!("Expected identifier, got {next:#?}"),
                ));
            }
        };

        let next = value.first();
        if next != Some(&Token::Keyword(Keywords::LeftArrow)) {
            return Ok(Arg::Simple(name));
        }
        value.pop_front();

        let arg_type = error!(Type::try_from(&mut *value), "Arg type")?;

        Ok(Arg::Named(name, arg_type))
    }
}

#[derive(Debug)]
pub struct ArgsTyped {
    lifetimes: Vec<String>,
    generics: Vec<Generic>,
    selft: Option<Type>,
    args: Vec<(String, Type)>,
}

impl TryFrom<&mut Parser> for ArgsTyped {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let mut args = vec![];
        let mut generics = vec![];
        let mut lifetimes = vec![];

        let next = value.pop_front_err("ArgsTyped")?;
        if next != Token::ParenOpen {
            return Err(error!(
                "ArgsTyped",
                format!("Expected ParenOpen, got {next:#?}"),
            ));
        }

        loop {
            let peek = value
                .first()
                .ok_or(error!("ArgsTyped", format!("Expected more tokens")))?;

            match peek {
                Token::Char(':') => {
                    generics.push(error!(Generic::try_from(&mut *value), "ArgsTyped")?)
                }
                Token::BackTick => {
                    value.pop_front();
                    match value.pop_front_err("ArgsTyped")? {
                        Token::Identifier(iden) => lifetimes.push(iden),
                        token => {
                            return Err(error!(
                                "ArgsTyped",
                                format!("Expected identifier, got {token:#?}")
                            ))
                        }
                    }
                }
                _ => break,
            }
        }

        let mut selft = None;

        loop {
            if value.first() == Some(&Token::ParenClose) {
                value.pop_front();
                break;
            }
            let arg = error!(Arg::try_from(&mut *value), "ArgsTyped")?;
            match arg {
                Arg::Generic(_) => return Err(error!("ArgsTyped", format!("Expected a named arg, got a generic, those should be defined before named args"))),
                Arg::Simple(_) => {
                    return Err(error!(
                        "ArgsTyped",
                        format!("Expected named arg, got a simple arg"),
                    ))
                }
                Arg::SelfA(selft2) => {
                    selft = Some(selft2)
                },
                Arg::Named(name, arg_type) => args.push((name, arg_type)),
            }
        }

        Ok(ArgsTyped {
            lifetimes,
            selft,
            generics,
            args,
        })
    }
}

impl ToString for ArgsTyped {
    fn to_string(&self) -> String {
        format!(
            "{}({}{})",
            if !self.generics.is_empty() {
                format!(
                    "<{}>",
                    &self.generics.iter().fold(String::new(), |str, generic| {
                        format!("{str}, {}", generic.to_string())
                    })[2..]
                )
            } else {
                format!("")
            },
            if let Some(selft) = &self.selft {
                format!(
                    "{}{}",
                    selft.to_string(),
                    if self.args.is_empty() {
                        format!("")
                    } else {
                        format!(", ")
                    }
                )
            } else {
                format!("")
            },
            if !self.args.is_empty() {
                (&self.args.iter().fold(String::new(), |str, arg| {
                    format!("{str}, {}: {}", arg.0, arg.1.to_string())
                })[2..])
                    .to_string()
            } else {
                format!("")
            }
        )
    }
}

#[derive(Debug)]
pub struct Args(Vec<String>);

impl TryFrom<&mut Parser> for Args {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let next = value.pop_front_err("Args")?;
        if next != Token::ParenOpen {
            return Err(error!("Args", format!("Expected ParenOpen, got {next:#?}"),));
        }

        let mut args = vec![];
        loop {
            if value.first() == Some(&Token::ParenClose) {
                value.pop_front();
                break;
            }

            let arg = error!(Arg::try_from(&mut *value), "Args")?;
            match arg {
                Arg::Simple(name) => args.push(name),
                _ => return Err(error!("Args arg", format!("Expected just simple args"),)),
            }
        }

        Ok(Self(args))
    }
}

impl ToString for Args {
    fn to_string(&self) -> String {
        format!(
            "|{}|",
            &self
                .0
                .iter()
                .fold(String::new(), |str, arg| { format!("{str}, {arg}") })[2..]
        )
    }
}
