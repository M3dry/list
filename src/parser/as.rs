use crate::tokenizer::{Keywords, Token};

use super::{
    error,
    exp::Exp,
    r#type::Type,
    Parser, ParserError, ParserErrorStack,
};

#[derive(Debug)]
pub struct As {
    exp: Exp,
    to: Type,
}
impl TryFrom<&mut Parser> for As {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let next = value.pop_front_err("As")?;
        if next != Token::ParenOpen {
            return Err(error!("As", format!("Expected parenOpen, got {next:#?}")));
        }
        let next = value.pop_front_err("As")?;
        if next != Token::Keyword(Keywords::As) {
            return Err(error!("As", format!("Expected as, got {next:#?}")));
        }

        let exp = error!(Exp::try_from(&mut *value), "Exp")?;
        let to = error!(Type::try_from(&mut *value), "Exp")?;

        let next = value.pop_front_err("As")?;
        if next != Token::ParenClose {
            return Err(error!("As", format!("Expected parenClose, got {next:#?}")));
        }

        Ok(Self { exp, to })
    }
}

impl ToString for As {
    fn to_string(&self) -> String {
        format!("({} as {})", self.exp.to_string(), self.to.to_string())
    }
}
