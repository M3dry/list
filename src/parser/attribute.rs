use crate::tokenizer::Token;

use super::{error, exp::Exp, Parser, ParserError, ParserErrorStack};

#[derive(Debug)]
pub enum Attribute {
    Inner(AttributeOps),
    Outer(AttributeOps),
}

impl TryFrom<&mut Parser> for Attribute {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let _ = error!("Attribute", value.pop_front(), [Token::Char('#')])?;
        let next = error!(
            "Attribute",
            value.pop_front(),
            [Token::Char('!'), Token::BracketOpen]
        )?;
        let inner = if next == Token::Char('!') {
            let _ = error!("Attribute", value.pop_front(), [Token::BracketOpen])?;
            true
        } else {
            false
        };

        let attr_ops = error!(AttributeOps::try_from(&mut *value), "Attribute")?;
        let _ = error!("Attribute", value.pop_front(), [Token::BracketClose])?;

        Ok(if inner {
            Self::Inner(attr_ops)
        } else {
            Self::Outer(attr_ops)
        })
    }
}

impl ToString for Attribute {
    fn to_string(&self) -> String {
        match self {
            Attribute::Inner(attr_ops) => format!("#![{}]", attr_ops.to_string()),
            Attribute::Outer(attr_ops) => format!("#[{}]", attr_ops.to_string()),
        }
    }
}

#[derive(Debug)]
pub enum AttributeOps {
    Command(String, Vec<AttributeOps>),
    Assignment(String, Exp),
    Identifier(String),
}

impl TryFrom<&mut Parser> for AttributeOps {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let next = error!(
            "AttributeOps",
            value.pop_front(),
            [Token::ParenOpen, Token::Identifier(_)]
        )?;
        Ok(match next {
            Token::ParenOpen => match error!(
                "AttributeOps",
                value.pop_front(),
                [Token::Identifier(_), Token::Char('=')]
            )? {
                Token::Identifier(iden) => {
                    let mut args = vec![];

                    loop {
                        let peek = value.first_err("AttributeOps")?;

                        if peek == &Token::ParenClose {
                            value.pop_front();
                            break Self::Command(iden, args);
                        }

                        args.push(error!(AttributeOps::try_from(&mut *value), "AttributeOps")?)
                    }
                }
                Token::Char('=') => {
                    let iden = error!("AttributeOps", value);
                    let exp = error!(Exp::try_from(&mut *value), "AttributeOps")?;
                    let _ = error!("AttributeOps", value.pop_front(), [Token::ParenClose])?;

                    Self::Assignment(iden, exp)
                }
                _ => unreachable!(),
            },
            Token::Identifier(iden) => Self::Identifier(iden),
            _ => unreachable!(),
        })
    }
}

impl ToString for AttributeOps {
    fn to_string(&self) -> String {
        match self {
            Self::Command(name, args) => format!(
                "{name}({})",
                &if args.is_empty() {
                    format!(", ")
                } else {
                    args.iter().fold(String::new(), |str, arg| {
                        format!("{str}, {}", arg.to_string())
                    })
                }[2..]
            ),
            Self::Assignment(var, exp) => format!("{var} = {}", exp.to_string()),
            Self::Identifier(var) => var.to_string(),
        }
    }
}
