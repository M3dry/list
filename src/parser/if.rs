use crate::tokenizer::{Keywords, Token};

use super::{error, exp::Exp, Parser, ParserError, ParserErrorStack};
#[derive(Debug)]
pub struct If {
    condition: Exp,
    true_branch: Exp,
    elif_branch: Vec<(Exp, Exp)>,
    false_branch: Exp,
}

impl TryFrom<&mut Parser> for If {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let next = value.pop_front_err("If")?;
        if next != Token::ParenOpen {
            return Err(error!("If", format!("Expected ParenOpen, got {next:#?}"),));
        }

        let next = value.pop_front_err("If")?;
        if next != Token::Keyword(Keywords::If) {
            return Err(error!("If", format!("Expected If keyword, got {next:#?}"),));
        }

        let condition = error!(Exp::try_from(&mut *value), "If")?;
        let true_branch = error!(Exp::try_from(&mut *value), "If")?;
        let mut elif_branch = vec![];

        loop {
            let peek = value.first();

            if peek == Some(&Token::Keyword(Keywords::Elif)) {
                value.pop_front();
                elif_branch.push((error!(Exp::try_from(&mut *value), "If")?, error!(Exp::try_from(&mut *value), "If")?))
            }

            break
        }

        let next = value.pop_front_err("If")?;
        if next != Token::Keyword(Keywords::Else) {
            return Err(error!("If", format!("Expected else, got {next:#?}")))
        }

        let false_branch = error!(Exp::try_from(&mut *value), "If")?;

        Ok(Self {
            condition,
            true_branch,
            elif_branch,
            false_branch,
        })
    }
}

impl ToString for If {
    fn to_string(&self) -> String {
        format!(
            "if {} {{{}}}{} else {{{}}}",
            self.condition.to_string(),
            self.true_branch.to_string(),
            self.elif_branch.iter().fold(String::new(), |str, elif| format!("{str} else if {} {{{}}}", elif.0.to_string(), elif.1.to_string())),
            self.false_branch.to_string()
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::snapshot;

    snapshot!(test_if, If::try_from, "if.lt");
    snapshot!(test_if_rust, If::try_from, "if.lt", rust);
}
