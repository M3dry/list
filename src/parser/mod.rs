pub mod args;
pub mod r#as;
pub mod attribute;
pub mod defun;
pub mod r#do;
pub mod r#enum;
pub mod exp;
pub mod file;
pub mod r#if;
pub mod r#impl;
pub mod lambda;
pub mod r#let;
pub mod r#match;
pub mod range;
pub mod r#struct;
pub mod r#type;
pub mod r#use;
pub mod r#trait;
pub mod module;

use std::{collections::VecDeque, path::PathBuf};

use std::ops::Index;

use crate::tokenizer::{Token, Tokens};

macro_rules! error {
    ($error:expr, $name:literal) => {
        $error.map_err(|mut err| {
            err.stack.push(ParserErrorStack {
                name: $name,
                file: file!(),
                location: (line!(), column!()),
            });
            err
        })
    };
    ($initial:expr, $err:expr$(,)?) => {
        ParserError {
            stack: vec![ParserErrorStack {
                name: $initial,
                file: file!(),
                location: (line!(), column!()),
            }],
            err: $err,
        }
    };
}

#[allow(unused_macros)]
macro_rules! snapshot {
    ($name:tt, $func:expr, $path:tt) => {
        snapshot!(
            $name,
            |parser| format!("{:#?}", $func(parser)),
            $path,
            "../../testdata/parser/"
        );
    };
    ($name:tt, $func:expr, $path:tt, rust) => {
        snapshot!(
            $name,
            |parser| match $func(parser) {
                Ok(res) => res.to_string(),
                Err(err) => format!("{err:#?}"),
            },
            $path,
            "../../testdata/rust/"
        );
    };
    ($name:tt, $func:expr, $path:tt, $out:literal) => {
        #[test]
        fn $name() {
            use crate::parser::Parser;

            let contents = include_str!(concat!("../../testdata/input/", $path));
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path($out);
            settings.bind(|| {
                insta::assert_snapshot!(contents
                    .lines()
                    .filter_map(|line| if line != "" {
                        Some(format!(
                            "{line}\n{}",
                            $func(&mut Parser::new(line.parse().unwrap()))
                        ))
                    } else {
                        None
                    })
                    .collect::<Vec<String>>()
                    .join("\n\n"));
            });
        }
    };
}

pub(crate) use error;
#[allow(unused_imports)]
pub(crate) use snapshot;

use self::file::File;

#[derive(Debug)]
pub struct ParserError {
    stack: Vec<ParserErrorStack>,
    err: String,
}

impl std::fmt::Display for ParserError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:", self.err)?;
        for stack in &self.stack {
            write!(f, "{stack}")?;
        }

        Ok(())
    }
}

#[derive(Debug)]
pub(crate) struct ParserErrorStack {
    name: &'static str,
    #[allow(dead_code)]
    file: &'static str,
    location: (u32, u32),
}

impl std::fmt::Display for ParserErrorStack {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}({};{})", self.name, self.location.0, self.location.0)
    }
}

#[derive(Debug)]
pub struct Parser {
    tokens: VecDeque<Token>,
}

impl Parser {
    pub fn new(tokens: Tokens) -> Self {
        Self {
            tokens: VecDeque::from(tokens.0),
        }
    }

    fn first(&self) -> Option<&Token> {
        if self.tokens.len() > 0 {
            Some(self.tokens.index(0))
        } else {
            None
        }
    }

    fn nth(&mut self, nth: usize) -> Option<&Token> {
        if self.tokens.len() > nth {
            Some(self.tokens.index(nth))
        } else {
            None
        }
    }

    fn pop_front_err(&mut self, func: &'static str) -> Result<Token, ParserError> {
        self.tokens
            .pop_front()
            .ok_or(error!(func, "Expected more tokens".to_string()))
    }

    fn pop_front(&mut self) -> Option<Token> {
        self.tokens.pop_front()
    }
}

pub(crate) fn from_file(path: &PathBuf) -> Result<String, ParserError> {
    File::try_from(&mut Parser::new(
        std::fs::read_to_string(&path)
            .map_err(|err| {
                error!(
                    "from_file",
                    format!("Got an err while opening the file at {path:#?}: {err:#?}")
                )
            })?
            .parse()
            .map_err(|err| {
                error!(
                    "from_file",
                    format!("Something went wrong while tokenizing file at {path:#?}: {err:#?}")
                )
            })?,
    ))
    .map(|f| f.to_string())
}
