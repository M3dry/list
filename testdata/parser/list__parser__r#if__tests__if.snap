---
source: src/parser/if.rs
expression: "contents.lines().filter_map(|line|\n                if line != \"\" {\n                        Some(format!(\"{line}\\n{}\",\n                                |parser| format!(\"{:#?}\", If::try_from (parser))\n                                (& mut Parser :: new(line.parse().unwrap()))))\n                    } else { None }).collect::<Vec<String>>().join(\"\\n\\n\")"
---
(if true 10 else 4)
Ok(
    If {
        condition: Literal(
            Bool(
                true,
            ),
        ),
        true_branch: Literal(
            Int(
                Int(
                    false,
                    10,
                ),
            ),
        ),
        elif_branch: [],
        false_branch: Literal(
            Int(
                Int(
                    false,
                    4,
                ),
            ),
        ),
    },
)

(if (= 10 5) 10 else (func 4 10))
Ok(
    If {
        condition: Infix(
            Equality(
                Literal(
                    Int(
                        Int(
                            false,
                            10,
                        ),
                    ),
                ),
                Literal(
                    Int(
                        Int(
                            false,
                            5,
                        ),
                    ),
                ),
            ),
        ),
        true_branch: Literal(
            Int(
                Int(
                    false,
                    10,
                ),
            ),
        ),
        elif_branch: [],
        false_branch: FuncCall(
            Variable(
                "func",
            ),
            [
                Literal(
                    Int(
                        Int(
                            false,
                            4,
                        ),
                    ),
                ),
                Literal(
                    Int(
                        Int(
                            false,
                            10,
                        ),
                    ),
                ),
            ],
        ),
    },
)

(if (= (func 1 10) (xyzfunc 100 (+ 1 10))) 100 elif (> 10 4) 10 else (xyzfunc 100 200))
Ok(
    If {
        condition: Infix(
            Equality(
                FuncCall(
                    Variable(
                        "func",
                    ),
                    [
                        Literal(
                            Int(
                                Int(
                                    false,
                                    1,
                                ),
                            ),
                        ),
                        Literal(
                            Int(
                                Int(
                                    false,
                                    10,
                                ),
                            ),
                        ),
                    ],
                ),
                FuncCall(
                    Variable(
                        "xyzfunc",
                    ),
                    [
                        Literal(
                            Int(
                                Int(
                                    false,
                                    100,
                                ),
                            ),
                        ),
                        Infix(
                            Add(
                                Literal(
                                    Int(
                                        Int(
                                            false,
                                            1,
                                        ),
                                    ),
                                ),
                                Literal(
                                    Int(
                                        Int(
                                            false,
                                            10,
                                        ),
                                    ),
                                ),
                            ),
                        ),
                    ],
                ),
            ),
        ),
        true_branch: Literal(
            Int(
                Int(
                    false,
                    100,
                ),
            ),
        ),
        elif_branch: [
            (
                Infix(
                    Greater(
                        Literal(
                            Int(
                                Int(
                                    false,
                                    10,
                                ),
                            ),
                        ),
                        Literal(
                            Int(
                                Int(
                                    false,
                                    4,
                                ),
                            ),
                        ),
                    ),
                ),
                Literal(
                    Int(
                        Int(
                            false,
                            10,
                        ),
                    ),
                ),
            ),
        ],
        false_branch: FuncCall(
            Variable(
                "xyzfunc",
            ),
            [
                Literal(
                    Int(
                        Int(
                            false,
                            100,
                        ),
                    ),
                ),
                Literal(
                    Int(
                        Int(
                            false,
                            200,
                        ),
                    ),
                ),
            ],
        ),
    },
)
